{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv\n",
    "from datasets import load_dataset\n",
    "from langchain.vectorstores.astradb import AstraDB\n",
    "from langchain.embeddings import OpenAIEmbeddings\n",
    "from langchain.schema import Document\n",
    "from langchain.prompts import ChatPromptTemplate\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "from langchain.schema.output_parser import StrOutputParser\n",
    "from langchain.schema.runnable import RunnablePassthrough\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "ASTRA_DB_COLLECTION = \"philosophers\"\n",
    "ASTRA_DB_APPLICATION_TOKEN = os.environ.get(\"ASTRA_DB_APPLICATION_TOKEN\")\n",
    "ASTRA_DB_API_ENDPOINT = os.environ.get(\"ASTRA_DB_API_ENDPOINT\")\n",
    "OPEN_AI_API_KEY = os.environ.get(\"OPENAI_API_KEY\")\n",
    "\n",
    "LANGSMITH_API_KEY = os.environ.get(\"LANGSMITH_API_KEY\")\n",
    "\n",
    "os.environ[\"LANGCHAIN_API_KEY\"] = LANGSMITH_API_KEY\n",
    "os.environ[\"LANGCHAIN_TRACING_V2\"] = \"true\"\n",
    "os.environ[\"LANGCHAIN_ENDPOINT\"] = \"https://api.smith.langchain.com\"\n",
    "os.environ[\"LANGCHAIN_PROJECT\"] = \"default\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "embedding = OpenAIEmbeddings()\n",
    "vstore = AstraDB(\n",
    "    embedding=embedding,\n",
    "    collection_name=ASTRA_DB_COLLECTION,\n",
    "    token=ASTRA_DB_APPLICATION_TOKEN,\n",
    "    api_endpoint=ASTRA_DB_API_ENDPOINT,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An example entry:\n",
      "{'author': 'aristotle', 'quote': 'Love well, be loved and do something of value.', 'tags': 'love;ethics'}\n",
      "Quote count: 450\n"
     ]
    }
   ],
   "source": [
    "philo_dataset = load_dataset(\"datastax/philosopher-quotes\")[\"train\"]\n",
    "print(\"An example entry:\")\n",
    "print(philo_dataset[16])\n",
    "\n",
    "print(f\"Quote count: {len(philo_dataset)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Inserted 450 documents.\n"
     ]
    }
   ],
   "source": [
    "docs = []\n",
    "\n",
    "for entry in philo_dataset:\n",
    "    metadata = {\"author\": entry[\"author\"]}\n",
    "    if entry[\"tags\"]:\n",
    "        # Add metadata tags to the metadata dictionary\n",
    "        for tag in entry[\"tags\"].split(\";\"):\n",
    "            metadata[tag] = \"y\"\n",
    "    # Add a LangChain document with the quote and metadata tags\n",
    "    doc = Document(page_content=entry[\"quote\"], metadata=metadata)\n",
    "    docs.append(doc)\n",
    "\n",
    "inserted_ids = vstore.add_documents(docs)\n",
    "print(f\"\\nInserted {len(inserted_ids)} documents.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(vstore.astra_db.collection(ASTRA_DB_COLLECTION).find())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1836, in _configure\n",
      "    handler = LangChainTracer(project_name=tracer_project)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 96, in __init__\n",
      "    self.client = client or get_client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 56, in get_client\n",
      "    _CLIENT = Client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 293, in __init__\n",
      "    _validate_api_key_if_hosted(self.api_url, self.api_key)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 198, in _validate_api_key_if_hosted\n",
      "    raise ls_utils.LangSmithUserError(\n",
      "langsmith.utils.LangSmithUserError: API key must be provided when using hosted LangSmith API\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 1083, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 927, in format\n",
      "    return fmt.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 663, in format\n",
      "    record.message = record.getMessage()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 367, in getMessage\n",
      "    msg = msg % self.args\n",
      "TypeError: not all arguments converted during string formatting\n",
      "Call stack:\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 87, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/traitlets/config/application.py\", line 1077, in launch_instance\n",
      "    app.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelapp.py\", line 737, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/tornado/platform/asyncio.py\", line 199, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 601, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 1905, in _run_once\n",
      "    handle._run()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 524, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 513, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 418, in dispatch_shell\n",
      "    await result\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 758, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/ipkernel.py\", line 426, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3048, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3103, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/async_helpers.py\", line 129, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3308, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3490, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3550, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/c7/cwsd4r9s4lv4fmcsxhn68f2m0000gn/T/ipykernel_34454/1475117834.py\", line 19, in <module>\n",
      "    response = chain.invoke(\"In the given context, what subject are philosophers most concerned with?\")\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/base.py\", line 1418, in invoke\n",
      "    callback_manager = get_callback_manager_for_config(config)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/config.py\", line 363, in get_callback_manager_for_config\n",
      "    return CallbackManager.configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1292, in configure\n",
      "    return _configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1839, in _configure\n",
      "    logger.warning(\n",
      "Message: 'Unable to load requested LangChainTracer. To disable this warning, unset the LANGCHAIN_TRACING_V2 environment variables.'\n",
      "Arguments: (LangSmithUserError('API key must be provided when using hosted LangSmith API'),)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1836, in _configure\n",
      "    handler = LangChainTracer(project_name=tracer_project)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 96, in __init__\n",
      "    self.client = client or get_client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 56, in get_client\n",
      "    _CLIENT = Client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 293, in __init__\n",
      "    _validate_api_key_if_hosted(self.api_url, self.api_key)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 198, in _validate_api_key_if_hosted\n",
      "    raise ls_utils.LangSmithUserError(\n",
      "langsmith.utils.LangSmithUserError: API key must be provided when using hosted LangSmith API\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 1083, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 927, in format\n",
      "    return fmt.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 663, in format\n",
      "    record.message = record.getMessage()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 367, in getMessage\n",
      "    msg = msg % self.args\n",
      "TypeError: not all arguments converted during string formatting\n",
      "Call stack:\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 87, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/traitlets/config/application.py\", line 1077, in launch_instance\n",
      "    app.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelapp.py\", line 737, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/tornado/platform/asyncio.py\", line 199, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 601, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 1905, in _run_once\n",
      "    handle._run()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 524, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 513, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 418, in dispatch_shell\n",
      "    await result\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 758, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/ipkernel.py\", line 426, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3048, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3103, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/async_helpers.py\", line 129, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3308, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3490, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3550, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/c7/cwsd4r9s4lv4fmcsxhn68f2m0000gn/T/ipykernel_34454/1475117834.py\", line 19, in <module>\n",
      "    response = chain.invoke(\"In the given context, what subject are philosophers most concerned with?\")\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/base.py\", line 1427, in invoke\n",
      "    input = step.invoke(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/base.py\", line 1907, in invoke\n",
      "    callback_manager = CallbackManager.configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1292, in configure\n",
      "    return _configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1839, in _configure\n",
      "    logger.warning(\n",
      "Message: 'Unable to load requested LangChainTracer. To disable this warning, unset the LANGCHAIN_TRACING_V2 environment variables.'\n",
      "Arguments: (LangSmithUserError('API key must be provided when using hosted LangSmith API'),)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1836, in _configure\n",
      "    handler = LangChainTracer(project_name=tracer_project)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 96, in __init__\n",
      "    self.client = client or get_client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 56, in get_client\n",
      "    _CLIENT = Client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 293, in __init__\n",
      "    _validate_api_key_if_hosted(self.api_url, self.api_key)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 198, in _validate_api_key_if_hosted\n",
      "    raise ls_utils.LangSmithUserError(\n",
      "langsmith.utils.LangSmithUserError: API key must be provided when using hosted LangSmith API\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 1083, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 927, in format\n",
      "    return fmt.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 663, in format\n",
      "    record.message = record.getMessage()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 367, in getMessage\n",
      "    msg = msg % self.args\n",
      "TypeError: not all arguments converted during string formatting\n",
      "Call stack:\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/threading.py\", line 937, in _bootstrap\n",
      "    self._bootstrap_inner()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/concurrent/futures/thread.py\", line 83, in _worker\n",
      "    work_item.run()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/concurrent/futures/thread.py\", line 58, in run\n",
      "    result = self.fn(*self.args, **self.kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/retrievers.py\", line 112, in invoke\n",
      "    return self.get_relevant_documents(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/retrievers.py\", line 186, in get_relevant_documents\n",
      "    callback_manager = CallbackManager.configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1292, in configure\n",
      "    return _configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1839, in _configure\n",
      "    logger.warning(\n",
      "Message: 'Unable to load requested LangChainTracer. To disable this warning, unset the LANGCHAIN_TRACING_V2 environment variables.'\n",
      "Arguments: (LangSmithUserError('API key must be provided when using hosted LangSmith API'),)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1836, in _configure\n",
      "    handler = LangChainTracer(project_name=tracer_project)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 96, in __init__\n",
      "    self.client = client or get_client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 56, in get_client\n",
      "    _CLIENT = Client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 293, in __init__\n",
      "    _validate_api_key_if_hosted(self.api_url, self.api_key)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 198, in _validate_api_key_if_hosted\n",
      "    raise ls_utils.LangSmithUserError(\n",
      "langsmith.utils.LangSmithUserError: API key must be provided when using hosted LangSmith API\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 1083, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 927, in format\n",
      "    return fmt.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 663, in format\n",
      "    record.message = record.getMessage()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 367, in getMessage\n",
      "    msg = msg % self.args\n",
      "TypeError: not all arguments converted during string formatting\n",
      "Call stack:\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/threading.py\", line 937, in _bootstrap\n",
      "    self._bootstrap_inner()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/concurrent/futures/thread.py\", line 83, in _worker\n",
      "    work_item.run()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/concurrent/futures/thread.py\", line 58, in run\n",
      "    result = self.fn(*self.args, **self.kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/passthrough.py\", line 191, in invoke\n",
      "    return self._call_with_config(identity, input, config)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/base.py\", line 840, in _call_with_config\n",
      "    callback_manager = get_callback_manager_for_config(config)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/config.py\", line 363, in get_callback_manager_for_config\n",
      "    return CallbackManager.configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1292, in configure\n",
      "    return _configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1839, in _configure\n",
      "    logger.warning(\n",
      "Message: 'Unable to load requested LangChainTracer. To disable this warning, unset the LANGCHAIN_TRACING_V2 environment variables.'\n",
      "Arguments: (LangSmithUserError('API key must be provided when using hosted LangSmith API'),)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1836, in _configure\n",
      "    handler = LangChainTracer(project_name=tracer_project)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 96, in __init__\n",
      "    self.client = client or get_client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 56, in get_client\n",
      "    _CLIENT = Client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 293, in __init__\n",
      "    _validate_api_key_if_hosted(self.api_url, self.api_key)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 198, in _validate_api_key_if_hosted\n",
      "    raise ls_utils.LangSmithUserError(\n",
      "langsmith.utils.LangSmithUserError: API key must be provided when using hosted LangSmith API\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 1083, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 927, in format\n",
      "    return fmt.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 663, in format\n",
      "    record.message = record.getMessage()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 367, in getMessage\n",
      "    msg = msg % self.args\n",
      "TypeError: not all arguments converted during string formatting\n",
      "Call stack:\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 87, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/traitlets/config/application.py\", line 1077, in launch_instance\n",
      "    app.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelapp.py\", line 737, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/tornado/platform/asyncio.py\", line 199, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 601, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 1905, in _run_once\n",
      "    handle._run()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 524, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 513, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 418, in dispatch_shell\n",
      "    await result\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 758, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/ipkernel.py\", line 426, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3048, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3103, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/async_helpers.py\", line 129, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3308, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3490, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3550, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/c7/cwsd4r9s4lv4fmcsxhn68f2m0000gn/T/ipykernel_34454/1475117834.py\", line 19, in <module>\n",
      "    response = chain.invoke(\"In the given context, what subject are philosophers most concerned with?\")\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/base.py\", line 1427, in invoke\n",
      "    input = step.invoke(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/prompts/base.py\", line 73, in invoke\n",
      "    return self._call_with_config(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/base.py\", line 840, in _call_with_config\n",
      "    callback_manager = get_callback_manager_for_config(config)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/config.py\", line 363, in get_callback_manager_for_config\n",
      "    return CallbackManager.configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1292, in configure\n",
      "    return _configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1839, in _configure\n",
      "    logger.warning(\n",
      "Message: 'Unable to load requested LangChainTracer. To disable this warning, unset the LANGCHAIN_TRACING_V2 environment variables.'\n",
      "Arguments: (LangSmithUserError('API key must be provided when using hosted LangSmith API'),)\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1836, in _configure\n",
      "    handler = LangChainTracer(project_name=tracer_project)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 96, in __init__\n",
      "    self.client = client or get_client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 56, in get_client\n",
      "    _CLIENT = Client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 293, in __init__\n",
      "    _validate_api_key_if_hosted(self.api_url, self.api_key)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 198, in _validate_api_key_if_hosted\n",
      "    raise ls_utils.LangSmithUserError(\n",
      "langsmith.utils.LangSmithUserError: API key must be provided when using hosted LangSmith API\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 1083, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 927, in format\n",
      "    return fmt.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 663, in format\n",
      "    record.message = record.getMessage()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 367, in getMessage\n",
      "    msg = msg % self.args\n",
      "TypeError: not all arguments converted during string formatting\n",
      "Call stack:\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 87, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/traitlets/config/application.py\", line 1077, in launch_instance\n",
      "    app.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelapp.py\", line 737, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/tornado/platform/asyncio.py\", line 199, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 601, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 1905, in _run_once\n",
      "    handle._run()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 524, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 513, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 418, in dispatch_shell\n",
      "    await result\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 758, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/ipkernel.py\", line 426, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3048, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3103, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/async_helpers.py\", line 129, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3308, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3490, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3550, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/c7/cwsd4r9s4lv4fmcsxhn68f2m0000gn/T/ipykernel_34454/1475117834.py\", line 19, in <module>\n",
      "    response = chain.invoke(\"In the given context, what subject are philosophers most concerned with?\")\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/base.py\", line 1427, in invoke\n",
      "    input = step.invoke(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/language_models/chat_models.py\", line 145, in invoke\n",
      "    self.generate_prompt(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/language_models/chat_models.py\", line 466, in generate_prompt\n",
      "    return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/language_models/chat_models.py\", line 324, in generate\n",
      "    callback_manager = CallbackManager.configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1292, in configure\n",
      "    return _configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1839, in _configure\n",
      "    logger.warning(\n",
      "Message: 'Unable to load requested LangChainTracer. To disable this warning, unset the LANGCHAIN_TRACING_V2 environment variables.'\n",
      "Arguments: (LangSmithUserError('API key must be provided when using hosted LangSmith API'),)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The subject that philosophers are most concerned with in the given context is knowledge or truth.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1836, in _configure\n",
      "    handler = LangChainTracer(project_name=tracer_project)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 96, in __init__\n",
      "    self.client = client or get_client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/tracers/langchain.py\", line 56, in get_client\n",
      "    _CLIENT = Client()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 293, in __init__\n",
      "    _validate_api_key_if_hosted(self.api_url, self.api_key)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langsmith/client.py\", line 198, in _validate_api_key_if_hosted\n",
      "    raise ls_utils.LangSmithUserError(\n",
      "langsmith.utils.LangSmithUserError: API key must be provided when using hosted LangSmith API\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 1083, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 927, in format\n",
      "    return fmt.format(record)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 663, in format\n",
      "    record.message = record.getMessage()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/logging/__init__.py\", line 367, in getMessage\n",
      "    msg = msg % self.args\n",
      "TypeError: not all arguments converted during string formatting\n",
      "Call stack:\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/runpy.py\", line 87, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/traitlets/config/application.py\", line 1077, in launch_instance\n",
      "    app.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelapp.py\", line 737, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/tornado/platform/asyncio.py\", line 199, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 601, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/base_events.py\", line 1905, in _run_once\n",
      "    handle._run()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/asyncio/events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 524, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 513, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 418, in dispatch_shell\n",
      "    await result\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/kernelbase.py\", line 758, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/ipkernel.py\", line 426, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3048, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3103, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/async_helpers.py\", line 129, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3308, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3490, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3550, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/c7/cwsd4r9s4lv4fmcsxhn68f2m0000gn/T/ipykernel_34454/1475117834.py\", line 19, in <module>\n",
      "    response = chain.invoke(\"In the given context, what subject are philosophers most concerned with?\")\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/base.py\", line 1427, in invoke\n",
      "    input = step.invoke(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/output_parsers/base.py\", line 170, in invoke\n",
      "    return self._call_with_config(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/base.py\", line 840, in _call_with_config\n",
      "    callback_manager = get_callback_manager_for_config(config)\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/runnables/config.py\", line 363, in get_callback_manager_for_config\n",
      "    return CallbackManager.configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1292, in configure\n",
      "    return _configure(\n",
      "  File \"/Users/hiltonrosenfield/miniconda3/envs/env_ragstack/lib/python3.9/site-packages/langchain_core/callbacks/manager.py\", line 1839, in _configure\n",
      "    logger.warning(\n",
      "Message: 'Unable to load requested LangChainTracer. To disable this warning, unset the LANGCHAIN_TRACING_V2 environment variables.'\n",
      "Arguments: (LangSmithUserError('API key must be provided when using hosted LangSmith API'),)\n"
     ]
    }
   ],
   "source": [
    "retriever = vstore.as_retriever(search_kwargs={'k': 3})\n",
    "\n",
    "prompt_template = \"\"\"\n",
    "Answer the question based only on the supplied context. If you don't know the answer, say you don't know the answer.\n",
    "Context: {context}\n",
    "Question: {question}\n",
    "Your answer:\n",
    "\"\"\"\n",
    "prompt = ChatPromptTemplate.from_template(prompt_template)\n",
    "model = ChatOpenAI(openai_api_key=OPEN_AI_API_KEY)\n",
    "\n",
    "chain = (\n",
    "    {\"context\": retriever, \"question\": RunnablePassthrough()}\n",
    "    | prompt\n",
    "    | model\n",
    "    | StrOutputParser()\n",
    ")\n",
    "\n",
    "response = chain.invoke(\"In the given context, what subject are philosophers most concerned with?\")\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env_ragstack",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
